### ISSUE TITLE
assoc() broken when slots=True

### ISSUE THREAD

**Issue by wbolster** [hynek/attrs#83]
it seems to me `attrs.assoc()` is broken when using `slots=True`:

```python
import attr

@attr.s(slots=True)
class Foo:
    bar = attr.ib()
    baz = attr.ib()


a = Foo(1, 2)
b = attr.assoc(a, baz=3)
print(b)
```

results in

```
Traceback (most recent call last):
...
ValueError: baz is not an attrs attribute on <class '__main__.Foo'>.
```

when i also add `frozen=True`, like this:

```python
@attr.s(slots=True, frozen=True)
class Foo:
    ...
```

i get this:

```
Traceback (most recent call last):
...
attr.exceptions.FrozenInstanceError
```

**Comment by hynek** [comment]
I believe this is related to #82.  Slots break more things than I originally expected. :|

**Comment by Tinche** [comment]
Looks like `assoc` is using a deprecated API. :smile_cat:

**Comment by Tinche** [comment]
I can take this, I promised already to make the tests for this use Hypothesis. I'll test all combinations of slots and frozen.

**Comment by hynek** [comment]
I think part of this and #82 is to come up with a new API for `Class.attribute`.  Something like `attr.by_name(Class, "attribute")`?

**Comment by Tinche** [comment]
Would you mind if that was a separate issue? I'll add the tests so the issue becomes obvious, then fix assoc in a dumb and slow way. When you're happy with the API assoc can be ported to it.

**Comment by hynek** [comment]
sure

**Comment by hynek** [comment]
Fixed in 1897dfd.
